internal/kms/aws_sts_metadata.go

```diff
@@ -0,0 +1,237 @@
+
+	// awsRoleSessionName is the name of the role session to connect with aws STS.
+	awsRoleSessionName = "ceph-csi-aws-sts-metadata"
+
+	// awsMetadataDefaultSecretsName is the default name of the Kubernetes Secret
+	// that contains the credentials to access the Amazon KMS. The name of
+	// the Secret can be configured by setting the `kmsSecretName`
+	// option.
+	//
+	// #nosec:G101, value not credential, just references token.
+	awsSTSMetadataDefaultSecretsName = "ceph-csi-aws-credentials"
+
+	// awsSTSSecretNameKey is the key for the secret name in the config map.
+	awsSTSSecretNameKey = "secretName"
+
+	// The following options are part of the Kubernetes Secrets.
+	//
+	// #nosec:G101.
+	awsSTSRoleARNKey = "awsRoleARN"
+	awsSTSCMKARNKey  = "awsCMKARN"
+	awsSTSRegionKey  = "awsRegion"
+
+	// tokenFilePath is the path to the file containing the OIDC token.
+	//
+	// #nosec:G101, value not credential, just path to the token.
+	tokenFilePath = "/var/run/secrets/tokens/oidc-token"
+)
+
+var _ = RegisterProvider(Provider{
+	UniqueID:    kmsTypeAWSSTSMetadata,
+	Initializer: initAWSSTSMetadataKMS,
+})
+
+type awsSTSMetadataKMS struct {
+	awsMetadataKMS
+
+	// AWS STS configuration options
+	role string
+}
+
+func initAWSSTSMetadataKMS(args ProviderInitArgs) (EncryptionKMS, error) {
+	kms := &awsSTSMetadataKMS{
+		awsMetadataKMS: awsMetadataKMS{
+			namespace: args.Tenant,
+		},
+	}
+
+	// get secret name if set, else use default.
+	err := setConfigString(&kms.secretName, args.Config, awsSTSSecretNameKey)
+	if errors.Is(err, errConfigOptionInvalid) {
+		return nil, err
+	} else if errors.Is(err, errConfigOptionMissing) {
+		kms.secretName = awsSTSMetadataDefaultSecretsName
+	}
+
+	// read the Kubernetes Secret with aws region, role & cmk ARN.
+	secrets, err := kms.getSecrets()
+	if err != nil {
+		return nil, fmt.Errorf("failed to get secrets: %w", err)
+	}
+
+	var found bool
+	kms.role, found = secrets[awsSTSRoleARNKey]
+	if !found {
+		return nil, fmt.Errorf("%w: %s", errConfigOptionMissing, awsSTSRoleARNKey)
+	}
+
+	kms.cmk, found = secrets[awsSTSCMKARNKey]
+	if !found {
+		return nil, fmt.Errorf("%w: %s", errConfigOptionMissing, awsSTSCMKARNKey)
+	}
+
+	kms.region, found = secrets[awsSTSRegionKey]
+	if !found {
+		return nil, fmt.Errorf("%w: %s", errConfigOptionMissing, awsSTSRegionKey)
+	}
+
+	return kms, nil
+}
+
+// getSecrets returns required STS configuration options from the Kubernetes Secret.
+func (as *awsSTSMetadataKMS) getSecrets() (map[string]string, error) {
+	c, err := k8s.NewK8sClient()
+	if err != nil {
+		return nil, fmt.Errorf("failed to connect to Kubernetes to "+
+			"get Secret %s/%s: %w", as.namespace, as.secretName, err)
+	}
+
+	secret, err := c.CoreV1().Secrets(as.namespace).Get(context.TODO(),
+		as.secretName, metav1.GetOptions{})
+	if err != nil {
+		return nil, fmt.Errorf("failed to get Secret %s/%s: %w",
+			as.namespace, as.secretName, err)
+	}
+
+	config := make(map[string]string)
+
```

Comment:
may be good to remove these blank lines.

Reply:
done